### DISCUSSION TITLE
Using bfs with fzf

### DISCUSSION THREAD

**Discussion by mathomp4** [tavianator/bfs#119] [General]
I'm a newbie to bfs and so am slowly learning how to use it. I've used `fd` a lot (it's replaced `find` for me in most cases), but I figure, hey, new-to-me tool!

So, one thing I thought I'd try is to maybe translate a couple FZF environment variables:
```
FZF_CTRL_T_COMMAND=fd --type file --color=always --no-ignore-vcs
FZF_ALT_C_COMMAND=fd --type d . --color=never --no-ignore-vcs
```
to `bfs` speak.

So far I have:
```
FZF_CTRL_T_COMMAND=bfs -type f -exclude -name .git
FZF_ALT_C_COMMAND=bfs -type d -exclude -name .git
```
which seems to work, so yay! But then I run `man bfs` and...hooboy. 😄 

I'm wondering if there are any of those other things in there I should look at and play with. Search strategy? Optimization?

**Comment by tavianator** [comment]
I have

```shell
export FZF_DEFAULT_OPTS='-x --ansi'
export FZF_CTRL_T_COMMAND="bfs -color -mindepth 1 -nohidden -printf '%P\n' 2>/dev/null"
export FZF_ALT_C_COMMAND="bfs -color -mindepth 1 -nohidden -type d -printf '%P\n' 2>/dev/null"
```

A breakdown:

- `-color`
  - Same as `fd`'s `--color=always`, needs `fzf --ansi`
- `-mindepth 1`
  - Skips the current directory, `.`
- `-nohidden`
  - Excludes hidden files/directories (same as `-exclude -hidden`)
- `-printf '%P\n'`
  - Prints the path without the leading `./`
- `2>/dev/null`
  - Hide errors

I also have

```shell
_fzf_compgen_path() {
    bfs -H "$1" -color -exclude \( -depth +0 -hidden \) 2>/dev/null
}

_fzf_compgen_dir() {
    bfs -H "$1" -color -exclude \( -depth +0 -hidden \) -type d 2>/dev/null
}
```

- `-H`
  - Follow symlinks in root paths (only), in case I do `link/**<TAB>`
- `-exclude \( -depth +0 -hidden \)`
  - Excludes hidden files unless they are root paths, in case I do `.git/**<TAB>`

**Comment by cstrahan** [comment]
Thanks @tavianator for `bfs` and sharing your configs!

I often want to hack on various dotfiles, so rather than using `-nohidden` I prefer to just exclude version control directories with `-exclude ( -name .git -or -name .hg )`.

So e.g. instead of this:

```bash
export FZF_CTRL_T_COMMAND="bfs -color -mindepth 1 -nohidden -printf '%P\n' 2>/dev/null"
```

I'd have:

```bash
export FZF_CTRL_T_COMMAND="bfs -color -mindepth 1 -exclude \\( -name .git -or -name .hg \\) -printf '%P\n' 2>/dev/null"
```

Just figured I'd share 😄.
